### DISCUSSION TITLE
My resource policy isn't applied

### DISCUSSION THREAD

**Discussion by markhowells** [serverless/serverless#10145] [Q&A] [answered]
I have a resource policy (grabbed directly from the docs at serverless.com) that doesn't appear to be associated with my lambda.

Here's the serverless.yml file

        # @format
        
        service: nhs-app-sender
        variablesResolutionMode: 20210326
        
        provider:
          name: aws
          runtime: nodejs14.x
          lambdaHashingVersion: 20201221
          logs:
            httpApi: true
          httpApi:
            metrics: true
            payload: "2.0"
          apiGateway:
            resourcePolicy:
              - Effect: Allow
                Principal: "*"
                Action: execute-api:Invoke
                Resource:
                  - execute-api:/*/*/*
                Condition:
                  IpAddress:
                    aws:SourceIp:
                      - "123.123.123.123"
          environment:
            keyID: test-1
        
        functions:
          loadTestData:
            handler: loadTestData.loadData
            vpc:
              securityGroupIds: ${ssm:/core/security-group/ecs}
              subnetIds: ${ssm:/core/private/subnetsAsList}
            events:
              - httpApi:
                  method: POST
                  path: /loaddata
                  authorizer:
                    type: aws_iam

There's no mention of the resource policy in either `cloudformation-template-create-stack.json` or `cloudformation-template-update-stack.json` 

I've spent a couple of days now googling like a donkey on drugs attempting to get it deployed but I can't for the life of me work out what I'm going wrong. I've enable/disabled the deprecation around `provider.apiGateway.resourcePolicy` and `provider.resourcePolicy` but nothing I do seems to have any effect. I'm sure I must be in error somewhere. 

Thanks, Mark

**Comment by pgrzesik** [comment]
Hello @markhowells - what exactly are you trying to do here? It looks like you're trying to set the `resourcePolicy` for API Gateway, not for the Lambda functions.

**Reply by markhowells** [reply]
@pgrzesik Well, I'm attempting to follow the model from [here](https://www.serverless.com/framework/docs/providers/aws/events/apigateway#resource-policy)

in turn, [this deprecation](https://www.serverless.com/framework/docs/deprecations#api-gateway-specific-configuration) relocates the policy to `provider.apiGateway.resourcePolicy`

However, I do indeed want to apply resourcePolicies to my functions...

**Reply by pgrzesik** [reply]
The resource policy on API Gateway is used to control access to the API itself, not to restrict functions themselves. As for function-specific permissions please consult the docs here: https://www.serverless.com/framework/docs/providers/aws/guide/iam

**Reply by markhowells** [reply]
Sorry, I may have mis-spoken. I want to control access TO my functions, not control what the functions can do.

I.e. I want to ensure that the Lambda's can only be accessed from one other service (The resourcePolicy I exemplified is just a test).

So, what I want to do is control access to the API itself. But when I specify as per my serverless.yml above, there's no mention of any RP other than the default...

**Comment by markhowells** [comment, marked as answer]
Update: I've just moved over to the older v1 REST api and I can now see a Resource Policy attached to the api gateway associated with my Lambda. The change I made was very small, simply what was required to go from httpApi to http events...

Not sure whether this is a bug or by design, but I cannot find any documentation to that effect...

**Reply by pgrzesik** [reply]
Hey, I've missed that you're using `httpApi` events - the `apiGateway` section under `provider` applies only to `http` events -> REST API APIGW integration. 

It's not a bug in any way, the `httpApi` event uses HTTP API APIGW and it does not support resource policies at all.

**Reply by markhowells** [reply]
Ah! OK. Is that by design or just not yet implemented? I have to say I didn't see that covered in the docs (although to be fair, the docs don't say that you _can_ use resource policies on httpApi either...).

Thanks for your help.

Mark

**Reply by pgrzesik** [reply]
Hard to say if it's by design or not implemented - it's not available on AWS side at all so we might get that in the future. I'm glad you were able to resolve your issue :raised_hands:
